Nitish add force parameter to create new section

hsitin opened the pull request.

hunner commented on lib/puppet/type/ini_setting.rb:
```diff
@@ -39,6 +39,11 @@ def munge_boolean_md5(value)
 
+  newparam(:force_new_section_creation, :boolean => true, :parent => Puppet::Parameter::Boolean) do
+    desc 'Create setting only if the section exists'
+    defaultto(true)
```
Defaulting to true seems like the opposite behavior of what currently exists?

hsitin replied:
The current behavior is that it creates the section and the setting if it does not exist in the ini file.  So defaulting force_new_section_creation to True will also create the new section.

hunner commented on lib/puppet/provider/ini_setting/ruby.rb:
```diff
@@ -35,7 +35,14 @@ def self.namevar(section_name, setting)
+        resource[:ensure] = :absent
+        resource[:force_new_section_creation]
+    else        
+      !resource[:force_new_section_creation]
```
Would this be idempotent? It seems like it would be trying to call create every puppet run.

hsitin commented:
1) Could you describe the purpose of this new parameter in your commit message or PR description.

The intend here is to manage the multiple redhat.repo config files on different servers. The files cannot be managed by Puppet as these will be generated by Satellite dynamically. The current behavior of the module creates new section and setting if it does not exist in the .ini file. What we want to achieve is that the module should not create new section in the .ini file if it does not exist. So we created a new param force_new_section_creation which is set to True by default, so that it doesn't break the previous logic. If it is set to False then it will not create the new section in the .ini file even if it is set to present.

2) Rubocop is failing on a few things. You can find these in the GitHub CI output
I have modified the code and fixed the errors. All checks have passed now.

3) The new parameter will need a readme entry before merging.
Added to the Readme file

pmcmaw commented:
@hunner are you happy enough with the changes?

david22swan commented:
@hunner ???

davidmalloncares commented:
merging, @hsitin - you might want to add some tests around the new behaviour though :)

mwhahaha commented:
@davidmalloncares @hsitin we're seeing failures with this change,

2018-06-22 14:59:50 | 2018-06-22 14:59:50,545 INFO: [1;31mError: /Stage[main]/Glance::Backend::Swift/Glance_swift_config[ref1/user]: Could not evaluate: Invalid parameter force_new_section_creation(:force_new_section_creation)[0m

Still investigating

davidmalloncares commented:
@mwhahaha np - let me know and we can pull the change back if required. @hsitin can you take a look to see if this has uncovered an issue?

mwhahaha commented on lib/puppet/provider/ini_setting/ruby.rb:
```diff
@@ -35,7 +35,14 @@ def self.namevar(section_name, setting)
-    !ini_file.get_value(section, setting).nil?
+    if ini_file.section?(section)
+      !ini_file.get_value(section, setting).nil?
+    elsif !resource[:force_new_section_creation]
```
So the issue i'm seeing is with this. If you're child object doesn't have this parameter (none of ours do because we're not inheirting from the upstram ini_setting type) this fails. I think what we need to do is check that this option is a value on the type before trying to consume it.

mwhahaha commented:
A proposed fix is available at https://github.com/puppetlabs/puppetlabs-inifile/pull/288 @hsitin and @davidmalloncares please review
